### ISSUE TITLE
Support Deferreds in Resource.render

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3711]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| esteve reported|
|-|-|
|Trac ID|trac#3711|
|Type|enhancement|
|Created|2009-03-30 17:04:18Z|
|Branch|https://github.com/twisted/twisted/tree/deferreds-in-resrender-3711|

Twisted.web2 supports returning Deferreds from the Resource#renderHTTP method, Twisted.web should support this same behavior in Resource.render.

Related: [#3621](https://github.com/twisted/twisted/issues/3621)

Attachments:

* [deferred-render.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/769/769c924dda383a378a63ab2433f72be8e17a0190/beafb25b6e6b7c7e0947e57006e66e1079b37a7f.patch) (10400 bytes) - added by esteve on 2009-03-30 17:06:46Z - This patch should let Resource return a Deferred, while maintaining compatibility
* [deferred-render-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/769/769c924dda383a378a63ab2433f72be8e17a0190/a1bcdf5da4c44330bc68fd0e0065b3fba4c09bdb.patch) (12602 bytes) - added by esteve on 2009-03-31 14:52:42Z - 
* [deferred-render-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/769/769c924dda383a378a63ab2433f72be8e17a0190/8bc8920b9567cfaa75c8363d45178515c11e5cd9.patch) (13913 bytes) - added by esteve on 2009-03-31 20:08:06Z - 
* [deferred_render-3711-landreville.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/769/769c924dda383a378a63ab2433f72be8e17a0190/dd010744df6ba4acd12ba26e8a8804a74bb1800c.patch) (8071 bytes) - added by Landreville on 2010-09-30 18:13:07Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3711 3711
type__enhancement enhancement
reporter__esteve esteve
priority__normal normal
milestone__ 
branch__deferreds_in_resrender_3711 deferreds-in-resrender-3711
branch_author__esteve__washort esteve, washort
status__new new
resolution__None None
component__web web
keywords__ 
time__1238432658000000 1238432658000000
changetime__1468585801665856 1468585801665856
version__None None
owner__hawkowl hawkowl
cc__thijs cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|esteve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set owner to esteve|
|-|-|

We'll need unit tests for any changes, can you create those as well?

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|esteve commented|
|-|-|

thijs: Here it is, I wrote it on the plane but forgot to upload it. Let me know if you need a more exhaustive test.

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|esteve commented|
|-|-|

I've updated the docs for IResource#render, but it probably needs a review. BTW, my patch makes trial complain about a lot of deprecated uses of NOT_DONE_YET, I can fix those, if you think deprecating NOT_DONE_YET is ok.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra removed owner|
|-|-|

Thanks esteve, putting it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [26597]) Branching to 'deferred-render-3711'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [26598]) Apply deferred-render-3.patch

refs [#3711](https://github.com/twisted/twisted/issues/3711)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to esteve|
|-|-|

Thanks for the patch Esteve!  I think you're on the right track.  Here are a few things, mostly minor:

  1. twisted/web/server.py
    1. The `render` method should have a docstring.  Since you're modifying it, you get to add one. ;)
    1. The existing code checks NOT_DONE_YET in two different ways - by identity and by equality.  Probably the new code should make this consistent, to avoid any really confusing cases where the two checks end up disagreeing (not sure if this is actually possible on current versions of CPython, but let's not risk it :).
    1. I think it's probably too early to deprecate NOT_DONE_YET.  This is a really old API with lots of users.  We should have a good replacement for it for a while, with docs helping people migrate, before we start yelling at them for not switching.
    1. in `_cbRender`, `types.StringType` should be replaced with `str` and probably `isinstance` should be used
    1. in `_ebRender`, the `r == UnsupportedMethod` check is redundant.  trap won't let that code execute if the check is going to fail.  Unless there's a subclass of `UnsupportedMethod` that should be handled differently?  But I don't think there is.
    1. in `_ebRender`, in the "HEAD" case, there is a second call to the resource's `render`.  This call isn't handling a Deferred result.  I think there's another ticket open about how this code is broken anyway, [#3684](https://github.com/twisted/twisted/issues/3684).
  1. twisted/web/test/test_web.py
    1. `SimpleDeferredResource` should have a docstring.  It should probably also not use `callLater`, but what exactly it _should_ use will depend on the next point...
    1. It doesn't seem as though anything tests the new Deferred support.  `test_simplestSiteDeferred` only does "resource traversal".  It doesn't invoke any render method.  So more tests are needed, I think.

The only really significant problem I see is the missing test coverage.  I suspect there are a few lurking corner cases that will need to be carefully tested to make sure they're working right and keep working right.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid<br>@dreid set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

This was a duplicate of [#11260](https://github.com/twisted/twisted/issues/11260), by the way.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_12"></a>|@mwhudson commented|
|-|-|

I think I've addressed most of the review comments for server.py now, apart from the one to do with HEAD handling (my brain hurts when I look at that code).

More tests still needed.

**Comment by twisted-trac** [comment]
|<img alt="Landreville's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|Landreville commented|
|-|-|

This has been sitting here for three months now. Would it be uncool if I kinda hijacked this ticket and posted a patch and test case that satisfies the problems in exarkuns review?

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_15"></a>|@dreid commented|
|-|-|

Absolutely not uncool.  In fact very very very cool.

**Comment by twisted-trac** [comment]
|<img alt="Landreville's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|Landreville removed owner<br>Landreville set status to `new`|
|-|-|

Ok I've attached a patch. I'm unsure of the test I added. It tests it alright, but it takes three seconds because I use deferLater to return a deferred from the resource and also deferLater to check if the request wrote the correct output.

The patch I submitted is deferred_render-3711-landreville.patch

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Is this patch against trunk or the branch?

**Comment by twisted-trac** [comment]
|<img alt="Landreville's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|Landreville commented|
|-|-|

trunk

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to Landreville|
|-|-|

I think there was a misunderstanding when you (Landreville) took over this ticket.  The latest patch seems to be a fresh start at implementing this feature, rather than building on the work in the branch linked in the ticket description (branches/deferred-render-3711).  Starting fresh can be a good thing, but only with a good reason.  The main reason not to start fresh is partially academic in this case at this point - it's almost always more work (both for authors and reviews) than finishing someone else's work.  Still, without a good reason, I'm not sure I want to spend the time to review this latest patch carefully when superficially looks like it's mostly a rehash of the work in branches/deferred-render-3711.  Ideally, a patch to finish this ticket would be made against the branch and just make the changes necessary to respond to the review comments on the ticket.

If you take a look at the unit tests in the branch, you'll see how they avoid a delay.

I hope this isn't too discouraging.  The help is really appreciated.  Sorry the way to proceed on this ticket wasn't more clear.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_21"></a>|@washort commented|
|-|-|

(In [31163]) Branching to "deferred-render-3711-2"

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_22"></a>|@washort commented|
|-|-|

(In [31182]) Branching to "deferred-render-3711-2".

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_23"></a>|@washort commented|
|-|-|

(In [31200]) Branching to 'deferred-render-3711-2'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_24"></a>|@washort removed owner|
|-|-|

I believe I've addressed the remaining issues in the review and improved test coverage sufficiently. Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve set owner to @washort|
|-|-|

1.

```
+            if body is NOT_DONE_YET:
+                # In a future release of Twisted, we should issue a
+                # DeprecationWarning here.
+                return
```

Please open a ticket and refer to the ticket number in the comment.

After thinking about it, maybe not? Aren't there use cases not covered by returning a `Deferred`?

2.

```
        self.method = "GET"                                                                    
        self._inFakeHead = True                                                                
        body = resrc.render(self)
```

It seems there may be a problem here if the render returns a `Deferred`?

3. It would be nice to use `self._errorResource` consistently.

4. There are 3 blank lines between `test_implicitHeadDeferred` and `test_badRenderResultDeferred` but only one between `test_badRenderResultDeferred` and `test_renderError`.

5. We should update the documentation, although I suppose it could be done in a future ticket. Please open it if you don't do it here.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_27"></a>|@lvh commented|
|-|-|

There's a problem with this ticket.

IResource.render says it can return a string or NOT_DONE_YET. Deferreds are neither. As a result, this would break the interface, and that would be bad.

The way to fix this is a new interface. See [#5152](https://github.com/twisted/twisted/issues/5152).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_28"></a>|@itamarst commented|
|-|-|

Adding another option to render() does not break the interface, it extends it. It's possible that extending the interface will break things, e.g. wrapper code that currently calls render() without expecting a Deferred. But extending the interface is not automatic grounds for rejection.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_29"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_28):
> Adding another option to render() does not break the interface, it extends it. It's possible that extending the interface will break things, e.g. wrapper code that currently calls render() without expecting a Deferred.

Perhaps "broken" is too vague a term (as is "adding another option", since Python functions don't have options, they have arguments and they have return values).  This does constitute an incompatible modification of the interface.  For example, if you have an interface that says it returns an `int` but then the interface were changed to be `int` or `str` or `None`, callers of that interface would be broken by that change; they would raise exceptions where they previously worked correctly.  The same is true of `int` versus "`Deferred` firing with `int`".

> But extending the interface is not automatic grounds for rejection.

I would say that extending it in this manner is, as it's a violation of the compatibility policy.  Of course, any such policy violation may be challenged and discussed on the mailing list, and if consensus can be achieved that nobody actually cares about the change and it won't break any running software, we can apply it as an exception to the policy.

It _might_ be technically compatible to change the documentation without actually changing any implementations or any callers of the interface, since the goal behind the compatibility policy is to avoid causing software to break, and a changed interface docstring wouldn't cause failures anywhere (not even in `verifyObject`, since it can't tell what return type we're talking about).  But that strikes me as rather pointless.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl set owner to @hawkowl|
|-|-|

I tried this on the plane. Lets see if the builders like it!
